src/auth.rs

```diff
@@ -72,25 +77,48 @@ pub fn get_hash<T: Digest>(text: &str) -> Vec<u8> {
     hasher.finalize().to_vec()
 }
 
-pub async fn handle_auth(req: ServiceRequest, cred: BasicAuth) -> Result<ServiceRequest> {
+/// When authentication succedes, return the request to be passed to downstream services.
+/// Otherwise, return an error response
+fn handle_auth(req: ServiceRequest) -> Result<ServiceRequest, ServiceResponse> {
     let (req, pl) = req.into_parts();
     let required_auth = &req.app_data::<crate::MiniserveConfig>().unwrap().auth;
 
-    if match_auth(cred.into(), required_auth) {
-        Ok(ServiceRequest::from_parts(req, pl).unwrap_or_else(|_| unreachable!()))
-    } else {
-        Err(HttpResponse::Unauthorized()
-            .header(
-                header::WWW_AUTHENTICATE,
-                header::HeaderValue::from_static("Basic realm=\"miniserve\""),
-            )
-            .body(build_unauthorized_response(
-                &req,
-                ContextualError::InvalidHttpCredentials,
-                true,
-                StatusCode::UNAUTHORIZED,
-            ))
-            .into())
+    if required_auth.is_empty() {
+        // auth is disabled by configuration
+        return Ok(ServiceRequest::from_parts(req, pl));
+    } else if let Ok(cred) = BasicAuthParams::try_from_request(&req) {
+        if match_auth(cred, required_auth) {
+            return Ok(ServiceRequest::from_parts(req, pl));
+        }
+    }
+
+    // auth failed; render and return the error response
+    let resp = HttpResponse::Unauthorized()
+        .append_header((
+            header::WWW_AUTHENTICATE,
+            header::HeaderValue::from_static("Basic realm=\"miniserve\""),
+        ))
+        .body(build_unauthorized_response(
+            &req,
+            ContextualError::InvalidHttpCredentials,
+            true,
+            StatusCode::UNAUTHORIZED,
+        ));
+
+    Err(ServiceResponse::new(req, resp))
+}
+
+pub fn auth_middleware<S>(
+    req: ServiceRequest,
+    srv: &S,
+) -> impl Future<Output = actix_web::Result<ServiceResponse>> + 'static
+where
+    S: Service<ServiceRequest, Response = ServiceResponse, Error = actix_web::Error>,
+    S::Future: 'static,
+{
+    match handle_auth(req) {
+        Ok(req) => Either::Left(srv.call(req)),
+        Err(resp) => Either::Right(ready(Ok(resp))),
```

Comment:
Withe the new error model in actix, we could no longer use `actix_web_httpauth::middleware::HttpAuthentication` to send send customized error reponse in case of failed authorization. Hence, it was neccessary to implement our own middleware to handle authentication and only use utility functions from `actix_web_httpauth` crate.

Reply:
I don't quite understand. actix_web_httpauth 0.6.0-beta.2 should be compatible with actix-web 4.0.0-beta.8 in my understanding. While your fix is acceptable, I wonder whether the compatibility couldn't be established here somehow.

Reply:
I think it is broken. In case of failed authorization we can  no longer send customized error response; only a basic one would be sent automatically.